src/Chapter3.hs

```diff
@@ -1006,6 +1139,64 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+class Contestant a where
+  fightIsOver :: a -> Bool -- fight is over if the Contestant is either dead or decided to run away. This results in a win for the other contestant
+  handleAttack :: a -> Attack -> a -- the other contestant attacks this contestant with a given Attack
+  applyNextAction :: (Contestant b) => a -> b -> (a, b)
+
+data Knight = Knight {
+  knightHealth :: Health,
+  knightDefense :: Defense,
+  knightActions :: [KnightAction]
+} deriving Show
+
+data KnightAction = AttackContestant Attack | DrinkAHealthPotion Health | CastASpell Defense deriving Show
+
+instance Contestant Knight where
+  fightIsOver knight = knightHealth knight <= Health 0
+  handleAttack knight attack = knight {knightHealth = calculateNewKnightHealth knight attack}
+  applyNextAction k@(Knight _ _ []) c = (k, c)
+  applyNextAction k@(Knight _ _  actions@(AttackContestant attack : _)) b = (k { knightActions = rotate actions}, handleAttack b attack)
+  applyNextAction k@(Knight health _ actions@(DrinkAHealthPotion moreHealth : _)) b = (k { knightHealth = health + moreHealth, knightActions = rotate actions}, b)
+  applyNextAction k@(Knight _ defense actions@(CastASpell moreDefense : _)) b = (k { knightDefense = defense + moreDefense, knightActions = rotate actions}, b)
+
+calculateNewKnightHealth :: Knight -> Attack -> Health
+calculateNewKnightHealth (Knight (Health health) (Defense defense) _) (Attack attack) = Health (health + defense - attack)
+
+data Monster = Monster {
+  monsterHealth :: Health,
+  monsterActions :: [MonsterAction]
+} deriving Show
+
+data MonsterAction = Fight Attack | RunAway deriving Show
+
+instance Contestant Monster where
+  fightIsOver (Monster _ (RunAway:_)) = True
+  fightIsOver monster = monsterHealth monster <= 0
+  handleAttack monster attack = monster {monsterHealth = calculateNewMonsterHealth monster attack}
+  applyNextAction m@(Monster _ []) c = (m, c)
+  applyNextAction m@(Monster _ (RunAway : _)) b = (m, b)
+  applyNextAction m@(Monster _ actions@(Fight attack : _)) b = (m { monsterActions = rotate actions}, handleAttack b attack)
+
+calculateNewMonsterHealth :: Monster -> Attack -> Health
+calculateNewMonsterHealth (Monster (Health health) _) (Attack attack) = Health (health - attack)
+
+data Result a b = Draw | AWon a | BWon b deriving Show
+
+fightContestants :: (Contestant a, Contestant b) => (a, b) -> Result a b
+fightContestants (a, b)
+  | fightIsOver a && fightIsOver b = Draw
+  | fightIsOver b = AWon a
+  | fightIsOver a = BWon b
+  | otherwise = if fightIsOver newB then AWon newA else fightContestants (newA', newB')
+  where
+    (newA, newB) = applyNextAction a b
+    (newB', newA') = applyNextAction newB newA
+
+rotate :: [a] -> [a]
+rotate as = (take len . drop 1 . cycle) as
+  where
+    len = length as
```

Comment:
This rotate function is a bit slow... For this particular case, it would be enough to just do:

```haskell
rotate :: [a] -> [a]
rotate l = drop 1 l ++ take 1 l
```

Though, since you're already familiar with `cycle`, there's no need to move elements from the beginning to the end. You can just call `cycle` once in the beginning and pattern match on the list head and tail.

Moreover, you can restrict the fact that the list is not-empty by using the `NonEmpty` data type:

* http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List-NonEmpty.html#t:NonEmpty

If you want, you can even model your types to make this case impossible. This pattern is called "Making illegal states unrepresentable". You can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

For example, if you really want to work with only infinite lists, you can create a separate data type for this:

```haskell
data Stream a = Stream a (Stream a)
```

and a helper function:

```haskell
cycleToStream :: NonEmpty a -> Stream a
```

Laziness in Haskell allows creating infinite data structures :exploding_head:

Reply:
That's exactly what I love about Haskell!

Reply:
I changed my implementation to use `NonEmpty`. I think the new `rotate` function is better in terms of performance. What do you think?